Bearded.Utilities/Algorithms/CoffmanGraham.cs

```diff
@@ -0,0 +1,190 @@
+                    return DecreasingNumberSequence.FromUnsortedNumbers(predecessorIndices);
+                }
+            }
+
+            private static ImmutableList<ImmutableHashSet<T>> createLayers(
+                // ReSharper disable once SuggestBaseTypeForParameter
+                IDirectedAcyclicGraph<T> graph, IList<T> ordering, int maxLayerSize)
+            {
+                var layersReversed = new List<ImmutableHashSet<T>.Builder>();
+                var elementToLayer = new Dictionary<T, int>();
+
+                for (var i = ordering.Count - 1; i >= 0; i--)
+                {
+                    // We fill in the layers back-to-front, always choosing the highest layer for the current element,
+                    // such that all its successors are in a higher layer, and the layer has less than W elements.
+                    // Since we are using a list where the right-most layer is 0, we are looking for the lowest k such
+                    // that (1) each successor is in a layer lower than k, and (2) layer k has less than W elements.
+
+                    // Requirement (1)
+                    var highestSuccessorLevel = graph.GetDirectSuccessorsOf(ordering[i])
+                        .Select(elmt => elementToLayer[elmt] + 1)
+                        .Aggregate(0, Math.Max);
+
+                    // Requirement (2)
+                    var candidateLayer = highestSuccessorLevel;
+                    while (
+                        candidateLayer < layersReversed.Count && layersReversed[candidateLayer].Count == maxLayerSize)
+                    {
+                        candidateLayer++;
+                    }
+
+                    // Expand the list as needed
+                    while (candidateLayer >= layersReversed.Count)
+                    {
+                        layersReversed.Add(ImmutableHashSet.CreateBuilder<T>());
+                    }
+
+                    // Add the element to the layer
+                    layersReversed[candidateLayer].Add(ordering[i]);
+                    elementToLayer.Add(ordering[i], candidateLayer);
+                }
+
+                return ImmutableList.CreateRange(layersReversed.Select(b => b.ToImmutable()).Reverse());
+            }
+        }
+
+        public static Instance<T> InstanceForGraph<T>(IDirectedAcyclicGraph<T> graph, int maxLayerSize)
+            where T : IEquatable<T>
+        {
+            return new Instance<T>(graph, maxLayerSize, isGraphReduced: false);
+        }
+
+        public static Instance<T> InstanceForAlreadyReducedGraph<T>(IDirectedAcyclicGraph<T> graph, int maxLayerSize)
+            where T : IEquatable<T>
+        {
+            return new Instance<T>(graph, maxLayerSize, isGraphReduced: true);
+        }
+
+        private struct DecreasingNumberSequence : IComparable<DecreasingNumberSequence>, IComparable
+        {
+            private readonly int[] numbers;
```

Comment:
Enlighten me... A list uses an array internally, so I am not sure what we would gain.

Reply:
I should have maybe commented below, in the static constructor methods. `ToArray` is doing an additional copy after essentially running `ToList` since the array can't be bigger than the number of elements, while the backing array of a list can be. That's a lot of extra copies that I don't think you need. Arrays aren't significantly faster than Lists.

Reply:
Size I care about the immutability here, changing it to an `ImmutableList`